pkg/kubelet/userns_manager.go

```diff
@@ -0,0 +1,476 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package kubelet
+
+import (
+	"encoding/json"
+	"fmt"
+	"math"
+	"os"
+	"path"
+	"sync"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	utilfeature "k8s.io/apiserver/pkg/util/feature"
+	runtimeapi "k8s.io/cri-api/pkg/apis/runtime/v1"
+	"k8s.io/kubernetes/pkg/features"
+)
+
+const bitsUint32 = 32
+
+type bitArray struct {
+	data       []uint32
+	firstIndex int
+	size       uint32
+}
+
+func makeBitArray(size uint32) *bitArray {
+	m := bitArray{
+		data:       make([]uint32, (size+bitsUint32-1)/bitsUint32),
+		firstIndex: 0,
+		size:       size,
+	}
+	return &m
+}
+
+func (b *bitArray) set(index uint32) {
+	b.data[index/bitsUint32] |= (uint32(1) << (index % bitsUint32))
+}
+
+func (b *bitArray) isSet(index uint32) bool {
+	return (b.data[index/bitsUint32]>>(index%bitsUint32))&0x1 == 1
+}
+
+func (b *bitArray) findAndSetFirstZero() (uint32, error) {
+	for i := b.firstIndex; i < len(b.data); i++ {
+		if b.data[i] == 0xFFFFFFFF {
+			continue
+		}
+		for j := uint32(0); j < bitsUint32; j++ {
+			if (b.data[i]>>j)&0x1 == 0 {
+				v := uint32(i)*bitsUint32 + j
+				b.set(v)
+				b.firstIndex = int(i)
+				return v, nil
+			}
+		}
+	}
+	return 0, fmt.Errorf("could not find an empty slot")
+}
+
+func (b *bitArray) clear(index uint32) {
+	i := index / bitsUint32
+	if i < uint32(b.firstIndex) {
+		b.firstIndex = int(i)
+	}
+	b.data[i] &= ^(1 << (index % bitsUint32))
+}
+
+// length for the user namespace to create (65536).
+const userNsLength = math.MaxUint16 + 1
```

Comment:
the reason why it is used is that originally 16 bits IDs were used (Linux supports 32 bits IDs since version 2.4), so the 16 bits have a meaning here.

Probably it is clearer if we write it as `(1 << 16)` so we don't need the `+ 1`

Reply:
Allocating exactly 65536 UIDs / GIDs to a container (or Pod) is more of a tradition than anything stronger, mind.

I get it, but we should explain in a comment why we're picking these numbers. If the KEP specified chunks of 100000 UIDs then that design would work just as well too.

Reply:
@sftim it is arguable if it is stronger than a tradition or not, but no point arguing about that.

Do you have any suggestion on what the comment should be here? I'm not sure I see what you want to make explicit.

Reply:
@sftim Can you provide a concrete suggestion of the comment you would like to see?